tools/clang/lib/Sema/SemaHLSLDiagnoseTU.cpp

```diff
@@ -592,14 +601,25 @@ void hlsl::DiagnoseTranslationUnit(clang::Sema *self) {
       hlsl::ShaderModel::GetByName(self->getLangOpts().HLSLProfile.c_str());
 
   llvm::SmallVector<VarDecl *, 16> GlobalsWithInit;
-  GatherGlobalsWithInitializers(self->getASTContext().getTranslationUnitDecl(),
-                                GlobalsWithInit);
-
+  llvm::SmallVector<VarDecl *, 16> SubObjects;
   std::set<FunctionDecl *> DiagnosedRecursiveDecls;
   llvm::SmallPtrSet<CallExpr *, 16> DiagnosedCalls;
   llvm::SmallPtrSet<DeclRefExpr *, 16> DeclAvailabilityChecked;
   llvm::SmallSet<SourceLocation, 16> DiagnosedTypeLocs;
 
+  GatherGlobalsWithInitializers(self->getASTContext().getTranslationUnitDecl(),
+                                GlobalsWithInit, SubObjects);
+
+  if (shaderModel->GetKind() == DXIL::ShaderKind::Library) {
+    DXIL::NodeLaunchType NodeLaunchTy = DXIL::NodeLaunchType::Invalid;
+    HLSLReachableDiagnoseVisitor Visitor(
+        self, shaderModel, shaderModel->GetKind(), NodeLaunchTy, nullptr,
+        DiagnosedCalls, DeclAvailabilityChecked, DiagnosedTypeLocs);
+    for (VarDecl *VD : SubObjects) {
+      Visitor.TraverseDecl(VD);
+    }
```

Comment:
nit: [brace style](https://llvm.org/docs/CodingStandards.html#don-t-use-braces-on-simple-single-statement-bodies-of-if-else-loop-statements).
```suggestion
    for (VarDecl *VD : SubObjects)
      Visitor.TraverseDecl(VD);
```